dev/benchmarks/macrobenchmarks/test/util.dart

```diff
@@ -0,0 +1,192 @@
+  Future<void> driverOps(WidgetController controller),
+  Future<void> setupOps(WidgetController controller),
+}) {
+  assert(() {
+    debugPrint(_kDebugWarning);
+    return true;
+  }());
+  final WidgetsBinding _binding = E2EWidgetsFlutterBinding.ensureInitialized();
+  assert(_binding is E2EWidgetsFlutterBinding);
+  final E2EWidgetsFlutterBinding binding = _binding as E2EWidgetsFlutterBinding;
+  binding.framePolicy = LiveTestWidgetsFlutterBindingFramePolicy.fullyLive;
+
+  testWidgets(testName, (WidgetTester tester) async {
+    assert((tester.binding as LiveTestWidgetsFlutterBinding).framePolicy ==
+        LiveTestWidgetsFlutterBindingFramePolicy.fullyLive);
+    app.main();
+    await tester.pumpAndSettle();
+
+    // The slight initial delay avoids starting the timing during a
+    // period of increased load on the device. Without this delay, the
+    // benchmark has greater noise.
+    // See: https://github.com/flutter/flutter/issues/19434
+    await tester.binding.delayed(const Duration(microseconds: 250));
+
+    final Finder scrollable =
+        find.byKey(const ValueKey<String>(kScrollableName));
+    expect(scrollable, findsOneWidget);
+    final Finder button =
+        find.byKey(ValueKey<String>(routeName), skipOffstage: false);
+    await tester.ensureVisible(button);
+    expect(button, findsOneWidget);
+    await tester.pumpAndSettle();
+    await tester.tap(button);
+
+    if (pageDelay != null) {
+      // Wait for the page to load
+      await tester.binding.delayed(pageDelay);
+    }
+
+    if (setupOps != null) {
+      await setupOps(tester);
+    }
+
+    await watchPerformance(binding, () async {
+      final Future<void> durationFuture = tester.binding.delayed(duration);
+      if (driverOps != null) {
+        await driverOps(tester);
+      }
+      await durationFuture;
+    });
+  }, semanticsEnabled: false, timeout: Timeout(timeout));
+}
+
+Future<void> watchPerformance(
+  E2EWidgetsFlutterBinding binding,
+  Future<void> action(),
+) async {
+  // This method might be good as part of e2e,
+  // so is the helper `PerformanceWatcher` class.
+  final FrameTimingSummarizer frameTimes = FrameTimingSummarizer();
+  final TimingsCallback watcher = frameTimes.addData;
+  binding.addTimingsCallback(watcher);
+  await action();
+  binding.removeTimingsCallback(watcher);
+  // TODO(CareF): determine if it's running on firebase and report metric online
+  binding.reportData = <String, dynamic>{'performance': frameTimes.summary};
+}
+
+class FrameTimingSummarizer {
+  FrameTimingSummarizer({this.data}) {
+    data ??= <FrameTiming>[];
+  }
+
+  List<FrameTiming> data;
+
+  void addData(List<FrameTiming> timings) {
+    data += timings;
+  }
+
+  Map<String, dynamic> get summary {
+    if (data.isEmpty) {
+      throw ArgumentError('durations is empty!');
+    }
+    final List<int> frameBuildTimeMicros = data.map<int>(
+      (FrameTiming datum) => datum.buildDuration.inMicroseconds,
+    ).toList();
```

Comment:
Then `kBuildBudget` would have to be a getter (or a const) so user won't expect to change it to change the related result.

Reply:
Done with the change, but let me know if `kBuildBudget` should still be a value that can be changed by customer, in which case it will need to be a member variable of the class.

Reply:
I realized I shouldn't do this because `data` is incremental.

Reply:
So it seems that I'll have to do the same as the `TimelineSummary` where multiple list generation and sort on the same object is performed... Or I'll have to make a cache system (I don't think I should...). Currently I reverted that moving to constructor change. Let me think about the trade-off.

Reply:
(moved to factory constructor)